BUG: Allow 'shape': () in __array_interface__ regardless of the data source

embray opened the pull request:
In [this comment](https://github.com/numpy/numpy/pull/6659#issuecomment-243188178) I mentioned a problem creating a type with an `__array_interface__` with `'shape': ()`.  This happens to work if either no `'data'` key is specified, or if `'data'` is a tuple (as in the case for actual `ndarray` instances).

However, if `'data'` is an object that exposes a buffer interface, it needlessly disallows zero-dim shapes.  This check was first added in #444, but was found in #2897 to be "unnecessary".  But #2897 only removed the check in one case, but not the other.  But my testing indicates that the check should not be necessary in either case.

It's possible there used to be bugs with this, but I don't think there are anymore.  In either case the dimensions (and ndim) are passed to `PyArray_NewFromDescr` which handles `ndim == 0` appropriately for array scalars.

embray committed:
BUG: Allow 'shape': () in __array_interface__ regardless what the data source is.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -6116,6 +6116,27 @@ def test_array_interface_itemsize():
+        __array_interface__ = interface1
+
+    interface2 = dict(interface1)
+    interface2['data'] = arr[0].tostring()
```
Might add a comment that strings support the buffer interface or something.

seberg replied:
Plus, I think the `tobytes` method looks better, not sure if we are deprecating the tostring one or not.

embray replied:
Sure, that all makes sense.

seberg commented:
LGTM, I am still a bit curious why that check used to be there, but I can't figure out how this can go bad (except the typical memory corruption that is always supported). One thing which could be cool, but independend, if we give a warning when the buffer seems too small for the array.

seberg commented:
Could you also reformat the commit message? The typical `BUG: ...` thing, not too long first line, etc. as described in the development workflow?

embray commented:
I can't figure out why the check used to be there either.  It was included as part of a fix to a mostly unrelated issue, and maybe at the time something else was bugging out if you tried to create an array scalar from a buffer.  I could probably go back in time and test it but it's probably not worth the bother.

embray commented:
Changed the title to read "BUG", actually, since this was more of a remnant misfeature as far as anyone can tell.

embray commented:
Updated to address requests, including using `tobytes()` instead of `tostring()` and added a little more of an explanatory note in the test.  Updated the commit message.

eric-wieser commented:
Was hoping that would restart the appveyor build

eric-wieser reviewed (approved):
LGTM

eric-wieser commented:
Thanks @embray!

embray commented:
Wow, cool.  I love it when >year old PRs that I forgot about get merged 🎊
